Add subnav capability

mattandrews opened the pull request:
This isn't quite ready to go yet until we finalise some of the other work going on, but still ready for code review.

At its most simple, this gives us the ability to add subnavigation to pages, like this:

![screen shot 2015-02-11 at 11 39 53](https://cloud.githubusercontent.com/assets/394376/6146251/a4e7ac7e-b1e2-11e4-9698-d9b9568eafbb.png)

At the same time it also fixes a possible regression in the existing nav:

#### Before:
![screen shot 2015-02-11 at 11 40 54](https://cloud.githubusercontent.com/assets/394376/6146272/eed5ead0-b1e2-11e4-8a10-6b1c496f7d08.png)

#### After:
![screen shot 2015-02-11 at 11 41 09](https://cloud.githubusercontent.com/assets/394376/6146267/e22ffe42-b1e2-11e4-8013-33788bac0701.png)

Also, this PR removes some of the nav options and adds them to a new, extended nav bar at the footer of the page:

![screen shot 2015-02-11 at 11 41 27](https://cloud.githubusercontent.com/assets/394376/6146285/0a7caf58-b1e3-11e4-834b-9c9f458665a1.png)

#### Caveats / concerns

The mobile view of this subnav isn't beautiful:
![screen shot 2015-02-11 at 11 45 02](https://cloud.githubusercontent.com/assets/394376/6146328/68246ed4-b1e3-11e4-8bcf-b009870e7de5.png)

... but on NGW, the subnav covers the main menu, which I don't think we'd want to do, so there's no clear path:

![screen shot 2015-02-11 at 11 45 47](https://cloud.githubusercontent.com/assets/394376/6146343/87388cd8-b1e3-11e4-9c93-555b74740a8e.png)

Finally, the CSS here isn't very beautiful either – the existing nav CSS is overly specific and not that modular and I didn't think it was worth a full-on refactor given that this work isn't even ready to deploy yet.

--

Thoughts and feedback encouraged.

mattandrews commented on frontend/app/model/Nav.scala:
```diff
@@ -0,0 +1,37 @@
+    NavItem("events", "/events", "Events"),
+    NavItem("masterclasses", "/masterclasses", "Masterclasses"),
+    NavItem("about", "/about", "About membership", Seq(
+        NavItem("supporters", "/supporters", "Supporters"),
```
(these endpoints don't exist yet, hence the [DO NOT MERGE] tag)

jennysivapalan replied:
Merged.

jennysivapalan replied:
:laughing: Sorry.

mattandrews replied:
:punc

mattandrews replied:
:facepunch:

davidrapson commented on frontend/app/views/fragments/nav.scala.html:
```diff
@@ -11,30 +12,46 @@
+
+
+
+@if(!isFooter) {
```
What's the thinking behind the `isFooter` logic?

mattandrews replied:
Basically: the subnav shouldn't show in the footer, and some of the top-level nav items should *only* show in the footer. Needed a way of telling the nav what state it was in.

davidrapson replied:
Gotcha: Not a huge fan of state/mode flags in fragments. If there's enough difference between the two it might be preferable to split them out into two fragments, I think it's easier to reason about even at the expense of some duplicated markup.

mattandrews replied:
Okay – that could work (there's a fair bit of footer logic around this file). @feedmypixel, anything to add?

feedmypixel replied:
separate fragments feels the way forward here and if there is anything inside of them you can fragment reuse do it, if not markup duplication for less "if statement-y" fragments seems sensible

davidrapson replied:
Matt and I chatted about this and we probably don't even need the footer nav, so no need for conditionals, but are going to wait and confirm with David M.

davidrapson commented on frontend/assets/stylesheets/components/nav/_nav.scss:
```diff
@@ -150,6 +151,60 @@ $global-nav-height: 36px;
     }
 }
 
+.nav--global--subnav {
```
Think we might be starting to get a BEM code-smell here. I'm not sure these need to be modifiers. Double modifier classes in BEM always feel like a red flag to me, but this is following what's already there so it's OK. Might be something to come back to as a later bit of refactoring once this goes out, doesn't feel like too big a deal yet.

My point is, what colour should we paint this bikeshed?

davidrapson commented on frontend/assets/stylesheets/components/nav/_nav.scss:
```diff
@@ -150,6 +151,60 @@ $global-nav-height: 36px;
+    margin-left: rem($gs-gutter);
+}
+
+.nav--global.nav--global--subnav.is-active {
```
Is this level of specificity needed?

davidrapson commented:
> Finally, the CSS here isn't very beautiful either – the existing nav CSS is overly specific and not that modular and I didn't think it was worth a full-on refactor given that this work isn't even ready to deploy yet.

Definitely agree with this. Needs a proper re-think, but not yet…

Possible option for mobile might be to lose the breadcrumb part and only show the sub-nav, not including that part in the toggle. Not particularly elegant though.

feedmypixel commented on frontend/app/model/Nav.scala:
```diff
@@ -0,0 +1,37 @@
+    NavItem("feedback", "/feedback", "Feedback", Nil, true)
+  )
+
+}
```
- could the details should be in config maybe?
- could you have a separate top and bottom nav list?

jennysivapalan commented on frontend/app/model/Nav.scala:
```diff
@@ -0,0 +1,37 @@
+  def fetchNav(url: String) = {
+    val navigationURLs = navigation.find { navItem =>
+      // we should return parent item if URL is a subnav
+      navItem.href == url || navItem.subNav.map(_.href).contains(url)
```
What a lovely bit of logic. I don't think it needs the comment.

mattandrews commented:
Now updated:

 * removed the footer nav
 * copied the display logic for the main nav (eg. no collapse, just sidescroll)
 * added subnav items to the main nav dropdown/hamburger

![screen shot 2015-02-24 at 16 31 08](https://cloud.githubusercontent.com/assets/394376/6353674/b299327e-bc42-11e4-9d44-735b8e5137ba.png)
![screen shot 2015-02-24 at 16 31 17](https://cloud.githubusercontent.com/assets/394376/6353675/b29d7456-bc42-11e4-9fe2-682725a5fa1d.png)

davidrapson commented:
:+1: LGTM, nice to get rid of the footer nav. As you mentioned we probably need to revisit the structure of the CSS here at some point, but this seems OK for now.

davidrapson commented on frontend/assets/stylesheets/components/nav/_nav.scss:
```diff
@@ -150,6 +156,59 @@ $global-nav-height: 36px;
+    }
+
+    .nav__item:after {
+        content: ">";
```
This is really minor, but this character for the divider looks a little odd. Maybe use the same  `›` character as .com?

mattandrews replied:
Done!

davidrapson replied:
:clap:

jennysivapalan commented:
In the interest of not looking like I'm targeting just @davidrapson ...merge or close this PR as it's pretty long running?

davidrapson commented:
@mattandrews Is this able to go out if we remove the sub nav items?

mattandrews commented:
Yep – will do so now. But @jamesoram needs to okay it too, bear with me...

davidrapson commented:
:bear:

mattandrews commented:
Okay – have removed the test subnav – @jamesoram (and @jennysivapalan), are you happy?

jennysivapalan commented:
:+1: :ship: :bulb:

mattandrews commented:
Let me know what you think @jamesoram!

jamesoram commented:
:+1:

davidrapson commented:
:shipit:
